cloudinit/sources/DataSourceVMware.py

```diff
@@ -292,6 +344,543 @@ def redact_keys(self):
+            if ud_file:
+                ud_path = os.path.join(get_vmware_imc_dir(), ud_file)
+                if not os.path.exists(ud_path):
+                    set_cust_error_status(
+                        "Error locating the cloud-init userdata file",
+                        "Userdata file is not found: %s" % ud_path,
+                        GuestCustEvent.GUESTCUST_EVENT_CUSTOMIZE_FAILED,
+                        cust_cfg,
+                    )
+                    return (None, None, None)
+                ud = util.load_file(ud_path).replace("\r", "")
+
+        set_customization_status(
+            GuestCustStateEnum.GUESTCUST_STATE_DONE,
+            GuestCustErrorEnum.GUESTCUST_ERROR_SUCCESS,
+        )
+        set_gc_status(cust_cfg, "Successful")
+        return (md, ud, vd)
+
+    def check_markers(self, cust_cfg):
+        product_marker = cust_cfg.marker_id
+        has_marker_file = check_marker_exists(
+            product_marker, os.path.join(self.paths.cloud_dir, "data")
+        )
+        return product_marker and not has_marker_file
+
+    def do_special_customization(self, cust_cfg, cust_cfg_dir):
+        is_pre_custom_successful = False
+        is_password_custom_successful = False
+        is_post_custom_successful = False
+        is_custom_script_enabled = False
+        custom_script = cust_cfg.custom_script_name
+        if custom_script:
+            is_custom_script_enabled = check_custom_script_enablement(cust_cfg)
+            if is_custom_script_enabled:
+                is_pre_custom_successful = do_pre_custom_script(
+                    cust_cfg, custom_script, cust_cfg_dir
+                )
+        is_password_custom_successful = do_password_customization(
+            cust_cfg, self.distro
+        )
+        if custom_script and is_custom_script_enabled:
+            ccScriptsDir = os.path.join(
+                self.paths.get_cpath("scripts"), "per-instance"
+            )
+            is_post_custom_successful = do_post_custom_script(
+                cust_cfg, custom_script, cust_cfg_dir, ccScriptsDir
+            )
+        if custom_script:
+            return (
+                is_pre_custom_successful
+                and is_password_custom_successful
+                and is_post_custom_successful
+            )
+        return is_password_custom_successful
+
+    def recheck_markers(self, cust_cfg):
+        product_marker = cust_cfg.marker_id
+        if product_marker:
+            cloud_dir = self.paths.cloud_dir
+            if not create_marker_file(cust_cfg, cloud_dir):
+                return False
+        return True
+
+
+def is_vmware_platform():
+    system_type = dmi.read_dmi_data("system-product-name")
+    if system_type is None:
+        LOG.debug("No system-product-name found")
+        return False
+    elif "vmware" not in system_type.lower():
+        LOG.debug("Not a VMware platform")
+        return False
+    return True
+
+
+def get_vmware_imc_dir():
```

Comment:
This should be a constant in this file near the top:

```python
#: Path to the VMware IMC directory.
IMC_DIR_PATH = "/var/run/vmware-imc"
```

Please note the constant is not prefixed with `VMWARE_` because that is implied given the datasource in which the constant is defined.

Reply:
Along with below imc specified functions, this one has been moved to guestcust_util.py, also added a constant there.